### DISCUSSION TITLE
Has anyone found how to create keys to be used for SSH (not FIDO2 resident keys)

### DISCUSSION THREAD

**Discussion by cpainchaud** [solokeys/solo2#156] [Q&A]
I am really struggling to generate & use 'classic' RSA/NISTp256 keys to use with SSH . I tried the PIV options from solo2-cli but it's quite empty there....

**Comment by msiqx9** [comment]
I saw a comment from Oliver Nelson 13 days ago on [indiegogo](https://www.indiegogo.com/projects/solo-v2-safety-net-against-phishing#/comments) which comments on SSH keys.  In case you can't find his comment, I'll post the full text below:

Got the keys I ordered/contributed to in March 2021. Firmware update on hacker key, normal one was already up-to-date. Works with ssh and the usual suspects (AWS, Google, Microsoft,…).
Hint: if you want to use it with your Mac, use homebrew to get a supported OpenSSH version (brew install openssh), then you can create a ssh key with the solo as additional auth like `ssh-keygen -t ed25519-sk`

**Reply by cpainchaud** [reply]
Hi,

 ed25519-sk are Fido2 resident keys which is what I'm trying to avoid because I have a doubt about their security based on several reading: the private key is stored in your hard drive or accessible on the token and full extractible. The token protects a sort of passphrase which should unlock the private key and I am trying to find understandable documentations which explain exactly how secure that is....

**Reply by jsonn** [reply]
The public key cryptography for `ed25519-sk` and `ecdsa-sk` is always done on the token and depends on the non-clonable on-chip secret. To access this key, a key handle is necessary. For non-resident keys, this is stored in the private "key" file and encrypted with the user passphrase. For resident keys, it can be queried directly from the token. So yes, resident keys are less secure than a non-resident key with a strong passphrase, but not for the reason given. You still need access to the token in either case.

**Reply by cpainchaud** [reply]
Thank you for your comment, I was not expecting to find a dramatic security issue there but I wanted to make sure I understand what was going on before using that technology. At first look, it was a big red flag when I read that the private key was stored on local computer or be publicly available when resident mode was chosen.

That said I found a few issues with daily usage of FIDo2 resident keys: none of the clients I used could tell me that were waiting for the key to be touched, you actually get a small time window to do it, too early or too late am and you will just spill random TOTP stings in currently active window. That is if you ever noticed you ssh/git clients were trying to authenticate.... From that point of view I like the smartcard approach better as the PIN requests are expected and managed by the clients

**Reply by jsonn** [reply]
I agree that at least with OpenSSH the UX can be improved. On Windows, you at least get the Hello prompt.

**Comment by jsonn** [comment]
Which SSH implementation are you trying to use? For OpenSSH, you need at least 8.2 to do what @msiqx9 mentioned. I haven't managed to get PuTTY to use a FIDO token directly, but I had success on Windows 10 and 11 using [SK-SSH-Agent](https://github.com/tetractic/SK-SSH-Agent/) as replacement for Pageant.

**Reply by cpainchaud** [reply]
Putty-Cac works like a charm, use 'puttyimp.exe' to import your Fido keys and their version of pageant
